### PR TITLE
Adding audience support

### PR DESCRIPTION
**Pull request by djw8605**

### CODE REVIEW

**Comment by djw8605** [comment]
This fixes #12

**Comment by bbockelm** [comment]
Please add the ability to do multiple audiences (i.e., accepts "T2_US_Nebraska" or "https://xrootd-local.unl.edu").

How did you test it?  I assume three tests are needed:
- How we handle the case where no audience is stated (outcome: accept)
- Case where audience matches (outcome: accept)
- Case where audience does not match (outcome: reject)

**Comment by jjg-123** [comment]
On 09/11/2018 03:00 PM, Brian Bockelman wrote:
>
> Please add the ability to do multiple audiences (i.e., accepts 
> "T2_US_Nebraska" or "https://xrootd-local.unl.edu").
>
If multiple audiences are specified, only one needs to match.
>
> How did you test it? I assume three tests are needed:
>
>   * How we handle the case where no audience is stated (outcome: accept)
>

I would vote that this should be an automatic reject instead. 
Audience=service that uses this token.  At least in the OAuth world, the 
first test is the audience and if the expected behavior is that if the 
consumer of the token does not find the right audience (itself), the 
token may be forged and should be rejected. Forged tokens are a real 
concern in OAuth and we should probably follow their lead in handling 
such things, since a SciToken is a type of access token too. That said, 
an audience may be general, such as "https://data.lsst.org" or even 
"LSST-datasets", meaning that you may have respectively what are 
correctly termed  resources (a specific (usually https) URL) or 
audiences (a moniker for a service). The plus with resources is that 
where the tokens may be used can be precisely described which is 
important for sensitive data.

**Comment by djw8605** [comment]
The JWT RFC explicitly says that `aud` is optional, I am very hesitant to deviate from that.

**Comment by djw8605** [comment]
> How we handle the case where no audience is stated (outcome: accept)

Yup

> Case where audience matches (outcome: accept)

Yup

> Case where audience does not match (outcome: reject)

Yup.  This also includes the case where the `aud` is not specified in the token, but is on the server.  And the opposite.

**Comment by jjg-123** [comment]
Hardly a starting point. Section 4.1 lists all the claims as optional 
and states:

"None of the claims
    defined below are intended to be mandatory to use or implement in all
    cases, but rather they provide a starting point for a set of useful,
    interoperable claims.  Applications using JWTs should define which
    specific claims they use and when they are required or optional. "

And the SciTokens claims document lists iss, exp and nbf as critical. My 
claim is simply that aud is sufficiently important to require it, just 
like these. Perhaps I am being simple-minded, but knowing who made it 
(iss), when it is valid (exp, nbf) and where it is intended to be used 
(aud) seem like pretty basic requirement for a self-describing security 
token...


Jeff

**Comment by bbockelm** [comment]
Hi,

Automatically rejecting would be a fairly significant break of the semantics for existing use cases.  We have a mechanism for this, however, that we've previously discussed at length - versioning.

If we really want to always include `aud`, then we can bump the version number and then change the rules accordingly.  Each implementation is supposed to know the minimum and maximum version number it can parse.

I'm fine with making `aud` mandatory for SciTokens, but not without using the existing mechanisms for doing such a change.

Brian

**Comment by jbasney** [comment]
[draft-ietf-oauth-jwt-bcp-03](https://tools.ietf.org/html/draft-ietf-oauth-jwt-bcp-03) contains some relevant guidance:

>   If the same issuer can issue JWTs that are intended for use by more
>   than one relying party or application, the JWT MUST contain an "aud"
>   (audience) claim that can be used to determine whether the JWT is
>   being used by an intended party or was substituted by an attacker at
>   an unintended party.  Furthermore, the relying party or application
>   MUST validate the audience value and if the audience value is not
>   present or not associated with the recipient, it MUST reject the JWT.

Just a draft at this point but it may be a useful reference on this and other JWT best practices.

I agree with Brian's comment about versioning. We don't want to break what's currently working for OSG.

**Comment by djw8605** [comment]
What is the conclusion on this?  I added multiple audience support.  Am I clear to merge?

**Note**: multiple audience support requires a relatively new version of the underlying JWT library, which is newer than the version distributed in EPEL.  So I will need to build a new version in the OSG.

File: src/scitokens_xrootd.py [hunk 1]
```diff
@@ -91,7 +94,17 @@ def config(fname):
...
         raise
+        
     for section in cp.sections():
+        if section.lower() == 'global':
+            if 'audience_json' in cp.options(section):
+                # Read in the audience as json.  Hopefully it's in list format or a string
+                g_audience = json.loads(cp.get("Global", "audience_json"))
+            elif 'audience' in cp.options(section):
+                g_audience = cp.get("Global", "audience")
+                if ',' in g_audience:
```
Comment (bbockelm) [inline src/scitokens_xrootd.py:107]: Can you always split the `g_global_audience`?  This way, `g_audience` is always a list and never a string.

**Review by bbockelm — CHANGES_REQUESTED** [review]
One small code change requested.

Are there any corresponding test cases?

File: src/scitokens_xrootd.py (hunk 1 above)
Comment (djw8605) [reply src/scitokens_xrootd.py:107]: Sure, I can always make `g_audience` a list, though the underlying library pyjwt doesn't really care.

**Comment by djw8605** [comment]
Just to let you know, I'm working on this.  Things that need to be done to get this test succesfully working:

1. Updated python-jwt built in OSG (https://opensciencegrid.atlassian.net/browse/SOFTWARE-3438)
2. Update to scitokens to support multiple-aud (https://github.com/scitokens/scitokens/pull/82)
3. Build scitokens in OSG from new release that includes multiple-aud

**Comment by djw8605** [comment]
Merging, tests pass.

**Pull request merged**